warehouse/query-core/src/main/java/datawave/next/DocIdIterator.java

```diff
@@ -0,0 +1,222 @@
+     *
+     * @param datatype
+     *            the datatype
+     * @return true if the datatype is accepted
+     */
+    protected boolean acceptDatatype(String datatype) {
+        if (datatypeFilter != null && !datatypeFilter.isEmpty() && !datatypeFilter.contains(datatype)) {
+            return false;
+        }
+
+        if (minDatatype != null && maxDatatype != null) {
+            return datatype.compareTo(minDatatype) >= 0 && datatype.compareTo(maxDatatype) <= 0;
+        }
+
+        return true;
+    }
+
+    @Override
+    protected void handleDatatypeMiss() {
+        stats.incrementDatatypeFilterMiss();
+        if (datatypeFilter != null) {
+            handleDatatypeFilterMiss(parser.getDatatype());
+        } else {
+            handleMinMaxDatatypeFilterMiss(parser.getDatatype());
+        }
+    }
+
+    /**
+     * Handle a datatype miss using the full set of requested datatypes
+     *
+     * @param datatype
+     *            the current datatype
+     */
+    protected void handleDatatypeFilterMiss(String datatype) {
+        String nextDatatype = datatypeFilter.higher(datatype);
+        if (nextDatatype == null) {
+            // rollover range?
+            Key start = new Key(row, prefix, value + '\u0000' + '\uffff');
+            Key stop = new Key(row, prefix, value + '\u0000' + '\uffff' + '\uffff');
+            Range range = new Range(start, false, stop, true);
+            safeSeek(range, true);
+            return; // no further datatypes exist, so we're done
+        }
+
+        // otherwise seek to the next datatype and continue iterating
+        Key start = new Key(row, prefix, value + '\u0000' + nextDatatype);
+        Key stop = new Key(row, prefix, value + '\u0000' + '\uffff');
+        Range range = new Range(start, false, stop, true);
+        safeSeek(range, true);
+    }
+
+    /**
+     * Handle a datatype miss using just a minimum and maximum datatype
+     *
+     * @param datatype
+     *            the current datatype
+     */
+    protected void handleMinMaxDatatypeFilterMiss(String datatype) {
+        if (datatype.compareTo(minDatatype) < 0) {
+            // seek to minimum
+            Key start = new Key(row, prefix, value + '\u0000' + minDatatype + '\u0000');
+            Key stop = new Key(row, prefix, value + '\u0000' + '\uffff');
+            Range range = new Range(start, false, stop, true);
+            safeSeek(range, true);
+        } else if (datatype.compareTo(maxDatatype) > 0) {
+            // rollover seek. For the EQ case generate a rollover range which causes hasNext() to be false
+            Key start = new Key(row, prefix, value + '\u0000' + maxDatatype + '\uffff');
+            Key stop = new Key(row, prefix, value + '\u0000' + '\uffff' + '\uffff');
+            Range range = new Range(start, false, stop, true);
+            safeSeek(range, true);
+        }
+    }
```

Comment:
pull the common code into methods

Reply:
Noted in https://github.com/NationalSecurityAgency/datawave/issues/2916